dask/array/routines.py

```diff
@@ -689,53 +716,82 @@ def histogram(a, bins=None, range=None, normed=False, weights=None, density=None
             "See the numpy.histogram docstring for more information."
         )
 
-    if not np.iterable(bins):
-        bin_token = bins
-        mn, mx = range
-        if mn == mx:
-            mn -= 0.5
-            mx += 0.5
-
-        bins = np.linspace(mn, mx, bins + 1, endpoint=True)
-    else:
-        bin_token = bins
-    token = tokenize(a, bin_token, range, weights, density)
+    if density and isinstance(bins, Array) and bins.ndim == 0:
+        raise NotImplementedError(
+            "When `density` is True, the number of bins cannot be a delayed Dask array. "
```

Comment:
The term `delayed Dask array` here confuses me. Do you mean a delayed scalar? We have no way to check if a `Delayed` object is a scalar or an array before computing, are there runtime checks for this if only delayed arrays are allowed?

Reply:
Yeah, I mean delayed scalar. Is that more consistent terminology? I wasn't sure what users would be more familiar with, since there's no type named "scalar". I just don't know how well-understood it is than an object of type `da.Array` might actually represent a scalar.

Reply:
Since it could be a `Delayed` object, the main restriction is that if it's a dask thing it must be a scalar dask thing. I'd write something like:

```suggestion
            "When `density` is True, `bins` cannot be a non-scalar Dask object. "
```

Reply:
I think you mean
```suggestion
            "When `density` is True, `bins` cannot be a scalar Dask object. "
```

It's okay if it's a non-scalar array (as long as it doesn't have NaN chunks).

Also worth noting this conditional doesn't check for scalar dask objects of other types. I'm not sure what that check should be. `or isinstance(bins, Delayed) and bins._length is not None and bins._length > 1`? What about dataframes?

Reply:
Ah, yeah, mistype.

> Also worth noting this conditional doesn't check for scalar dask objects of other types. I'm not sure what that check should be. or isinstance(bins, Delayed) and bins._length is not None and bins._length > 1? What about dataframes?

Is there a way to check at runtime that a nonscalar was used for bins, and just try to catch programmer error as best as possible at graph build time? I might only allow arrays and `Delayed` values and disallow other things like dataframes. There is no way to know what a `Delayed` value contains until runtime, so you'd have to have a runtime check for that.

Reply:
> Is there a way to check at runtime that a nonscalar was used for bins

We could add a task to the graph just to do some checking, but I figure NumPy will ultimately tell us at runtime if it can use the input or not. The error message will likely not be very helpful though. But replicating that logic—or catching the error and being able to tell users more meaningfully where they went wrong—could be complicated.

> I might only allow arrays and `Delayed` values and disallow other things like dataframes.

I went with this. _Of dask types_, only Array and Delayed are allowed (for the non-boolean parameters). I didn't make any effort to check non-Dask types, since without being overly restrictive, and duplicating NumPy code, it's kind of not possible.

Reply:
Out of curiosity, is it a goal for dask to catch all type / usage errors at graph construction time? Seems to me like the majority of the code (in Array at least) doesn't do much rigorous typechecking and would fail in all sorts of ways if you passed in the wrong things. Is that something you're trying to move away from, and get stricter? Seems to me like that would require a lot of work on foundational tooling, like decorators for typechecking. But it's kind of a losing battle in Python.